tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
     return new OperationIterator(this);
   }
 
+  private GraphOperation graphOp(Operand<?> operand) {
+    checkInput(operand);
+    return (GraphOperation) operand.op();
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Respects control dependencies.
+   *
+   * If both {@code allowedBodyOps} and {@code forbiddenBodyOps} are null, forbids 0-input ops in the body.
+   *
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.
+   * @param outputs the outputs of the subgraph
+   * @param allowedBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenBodyOps}) if null.
+   * @param forbiddenBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
```

Comment:
Should we have an overload for this that just accepts the first two arguments and calls this with `Collections.emptySet()` for those arguments? Or document that all the arguments must be non-null and expect our users to suppled `Collections.emptySet()` on their own?

Reply:
The reason I made it take nulls in the first place is because calling with both as emptySet errors whenever something is in the body (since we're allowing the empty set of ops).  I think an overload is a good idea though, and specifying the first two args must be non-null.